src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs

```diff
@@ -181,6 +189,34 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
             return locals?.ToImmutableAndFree() ?? ImmutableArray<LocalSymbol>.Empty;
         }
 
+        // When a VariableDeclaration is used in a deconstruction, there are two cases:
+        // - deconstruction is set, type may be set (for "var"), and no declarators. For instance, `var (x, ...)` or `(int x, ...)`.
+        // - deconstruction is null, type may be set, and there is one declarator holding the identifier. For instance, `int x` or `x`.
+        private void CollectLocalsFromDeconstruction(VariableDeclarationSyntax declaration, TypeSyntax varSyntax, ArrayBuilder<LocalSymbol> locals)
+        {
+            if (declaration.Deconstruction != null)
+            {
+                var nestedVarSyntax = varSyntax ?? declaration.Type;
+                foreach (var variable in declaration.Deconstruction.Variables)
+                {
+                    CollectLocalsFromDeconstruction(variable, nestedVarSyntax, locals);
+                }
+            }
+            else
+            {
+                Debug.Assert(declaration.Variables.Count == 1);
+                var declarator = declaration.Variables[0];
+
+                SourceLocalSymbol localSymbol = SourceLocalSymbol.MakeDeconstructionLocal(
+                                                            this.ContainingMemberOrLambda,
+                                                            this,
+                                                            declaration.Type,
+                                                            declarator.Identifier,
+                                                            isVar: varSyntax != null);
```

Comment:
The `var` syntax can come from multiple parents up. For example, for `y` in `var (x, (y, z))`. So that is why I prefer to pass it in.
Two examples:

1. In `var (x, (y, z))`, we create a local that may be implicitly typed for `y` because isVar will be true. That local will have its type syntax set to null.
2. In `(var x, int y)`, we should also create an implicitly typed local for `x`, with type syntax pointing to `var`. I think that's a bug there. I'll fix.